### PR TITLE
descartes_tests package

### PR DESCRIPTION
**Pull request by Jmeyer1292**
Following up on #198, this package:
 * Moves most of the unit tests to a new `descartes_tests` package
 * Removes weird exported test folders
 * Re-enables install space testing!

Thanks again to @130s and @ipa-mdl for their help and hard-work with CI. I really appreciate it. Sorry I'm slow as molasses fixing this stuff.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,8 +8,8 @@ notifications:
     on_failure: always
 
 env:
-  global:
-    - BEFORE_SCRIPT='catkin config -w $CATKIN_WORKSPACE --no-install'
+  # global:
```
Comment (shaun-edwards) [inline .travis.yml]: Can this be deleted now?

File: descartes/package.xml [hunk 2]
```diff
@@ -16,7 +16,7 @@
   <run_depend>descartes_planner</run_depend>
   <run_depend>descartes_trajectory</run_depend>
   <run_depend>descartes_utilities</run_depend>
-
+  <run_depend>descartes_tests</run_depend>
```
Comment (shaun-edwards) [inline descartes/package.xml]: `descartes` is a meta-package, correct?  Should we include a test package in the dependency list?  This will include it in the debian, even though none of the other packages really need it.  Mostly, curious.

File: descartes_tests/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+if(CATKIN_ENABLE_TESTING)
+  # Trajectory Unit Tests
+  catkin_add_gtest(${PROJECT_NAME}_trajectory_utest
+    test/trajectory/utest.cpp
+    test/trajectory/axial_symmetric_pt.cpp
+    test/trajectory/cart_trajectory_pt.cpp
+    test/trajectory/joint_trajectory_pt.cpp
+    test/trajectory/cartesian_robot_test.cpp
+  )
+  target_compile_definitions(${PROJECT_NAME}_trajectory_utest PUBLIC GTEST_USE_OWN_TR1_TUPLE=0)
```
Comment (shaun-edwards) [inline descartes_tests/CMakeLists.txt:58]: What does `PUBLIC GTEST_USE_OWN_TR1_TUPLE=0` do?

File: descartes_tests/package.xml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+  <maintainer email="jmeyer@swri.org">Jonathan Meyer</maintainer>
+  <license>Apache 2.0</license>
+
+  <test_depend>gtest</test_depend>
+  <buildtool_depend>catkin</buildtool_depend>
+  <build_depend>descartes_moveit</build_depend>
+  <build_depend>descartes_planner</build_depend>
+  <build_depend>descartes_trajectory</build_depend>
+  <run_depend>descartes_moveit</run_depend>
```
Comment (shaun-edwards) [inline descartes_tests/package.xml:18]: Should these be `test_depend`'s?

File: descartes_tests/test/trajectory/cart_trajectory_pt.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
+/*
```
Comment (shaun-edwards) [inline descartes_tests/test/trajectory/cart_trajectory_pt.cpp]: Is this a new file? With new tests?

File: descartes_trajectory/test/cart_trajectory_pt.cpp [hunk 6]
```diff
@@ -1,160 +0,0 @@
-/*
```
Comment (shaun-edwards) [inline descartes_trajectory/test/cart_trajectory_pt.cpp]: It doesn't look like all the tests in this file got captured in the new file above?

**Review by shaun-edwards — CHANGES_REQUESTED** [review]
Just a couple of questions/simple issues.  +1 for doing this!

File: .travis.yml (hunk 1 above)
Comment (Jmeyer1292) [reply .travis.yml]: +1

File: descartes/package.xml (hunk 2 above)
Comment (Jmeyer1292) [reply descartes/package.xml]: I was just trying to be diligent, and I have very little knowledge of the Debian release process. I'm happy to drop this line if doing so has no real consequence.

File: descartes_tests/CMakeLists.txt (hunk 3 above)
Comment (Jmeyer1292) [reply descartes_tests/CMakeLists.txt:58]: See docs [here](https://github.com/google/googletest/tree/master/googletest#choosing-a-tr1-tuple-library). Basically, I'm telling GTEST to use the c++11 `std::tuple`. A while ago (perhaps on Indigo?) I had trouble with a strange build error involving multiple versions of said class and this was the solution. This line was copied from the other tests.

File: descartes_tests/package.xml (hunk 4 above)
Comment (Jmeyer1292) [reply descartes_tests/package.xml:18]: Honestly, I don't know. It would seem in package format 2, they would be, but I'm unclear as to the original format. Perhaps I should put everything in the `CMakeLists.txt` file in a `if(testing)` block.

File: descartes_tests/test/trajectory/cart_trajectory_pt.cpp (hunk 5 above)
Comment (Jmeyer1292) [reply descartes_tests/test/trajectory/cart_trajectory_pt.cpp]: See below.

File: descartes_trajectory/test/cart_trajectory_pt.cpp (hunk 6 above)
Comment (Jmeyer1292) [reply descartes_trajectory/test/cart_trajectory_pt.cpp]: I deleted an extremely long running test (>1 minute), `getPoses` that counted angle discretization. A real pain to work with while doing this work. I will be-grudgingly put it back.

Comment (shaun-edwards) [reply descartes_trajectory/test/cart_trajectory_pt.cpp]: You know there is a way to disable a test, by adding DISABLED_ to the test name if I remember correctly.  I do this temporarily when I am debugging.

**Comment by mathias-luedtke** [comment]
Great job!
So you want to make `CartesianRobot` accessible to others?
This will implicitly answer the meta-package and dependency questions..

**Review by davetcoleman — APPROVED** [review]
Great contribution! Its not clear to me, though, the advantage of moving the unit tests farther from their corresponding source code. To me tests should in every package and provide coverage for that package. A descartes_tests package might have merit for integration tests, however, that require lots of special config files.

File: descartes_tests/package.xml (hunk 4 above)
Comment (davetcoleman) [reply descartes_tests/package.xml:18]: why not just switch to format 2?

**Comment by Jmeyer1292** [comment]
@davetcoleman If I remember correctly, the industrial_ci package made a change a while back that forced its users to only use the ROS install space, so the only way two packages could share resources was through install rules. This meant that to have meaningful Descartes tests (where one needs a sample robot model to test planners or points), I would need to install test artifacts (like a dummy cartesian robot) or create a separate package that hosted them all. I picked the latter option for some reason.

**Comment by 130s** [comment]
I also had a similar (but not the same) situation recently (in my case I used `PKGS_DOWNSTREAM` but this might not suit for this PR).
Besides, I opened https://github.com/ros-industrial/industrial_ci/issues/231 for further improvement for this kind of situation.

**Comment by davetcoleman** [comment]
> I would need to install test artifacts (like a dummy cartesian robot) or create a separate package that hosted them all. I picked the latter option for some reason.

You can still host your test data (similar to moveit_resources) in a test package, but have your tests located in their correct package location, right?

**Comment by Jmeyer1292** [comment]
> You can still host your test data (similar to moveit_resources) in a test package, but have your tests located in their correct package location, right?
I don't really understand how this all works, but it looks like moveit resources are config files. What I'm talking about are C++ libraries. I can't test planners without installing test classes from previous packages. 

Anyway, this whole test re-organization is just meant to ease the mental burden of pushing back into this repository. Wherever the tests are, they will still fail if something goes wrong. I will happily accept more pull requests to fix it how you want.
